csrc/kernel_ir.h

```diff
@@ -521,84 +519,64 @@ class MBarrierWait final : public Expr {
   }
 };
 
-// CpAsyncWait represents wait intrinsics for cp.async
-class CpAsyncWait final : public Expr {
+// AsyncWait represents wait intrinsics for cp.async, cp.async.bulk and
+// wgmma.mma_async
+class AsyncWait final : public Expr {
  public:
   using Expr::Expr;
 
-  explicit CpAsyncWait(IrBuilderPasskey passkey, int64_t keep_stages = 0);
+  explicit AsyncWait(
+      IrBuilderPasskey passkey,
+      AsyncOpType async_op_type,
+      int64_t keep_stages = 0);
 
   NVFUSER_DECLARE_CLONE_AND_CREATE
 
   const char* getOpString() const override {
-    return "CpAsyncWait";
+    return "AsyncWait";
   }
 
   std::string toString(int indent_size = 0) const override;
   std::string toInlineString(int indent_size = 0) const override;
 
+  const char* ptx() const;
+  bool memory() const;
+
+  AsyncOpType asyncOpType() const {
+    return attribute<AsyncOpType>(0);
+  }
+
   //! Returns the remaining number of stages that are not synchronized
   //!  after this op.
   int64_t keepStages() const {
-    return attribute<int64_t>(0);
+    return attribute<int64_t>(1);
   }
 };
 
-// CpAsyncCommit represents commit intrinsics for cp.async
+// AsyncCommit represents commit intrinsics for cp.async
 //  A commit intrinsic communicates delimiter of transaction groups
 // to the async load hardware. Example usage see [Cicular buffer].
-class CpAsyncCommit final : public Expr {
- public:
-  using Expr::Expr;
-
-  explicit CpAsyncCommit(IrBuilderPasskey passkey);
-
-  NVFUSER_DECLARE_CLONE_AND_CREATE
-
-  const char* getOpString() const override {
-    return "CpAsyncCommit";
-  }
-
-  std::string toString(int indent_size = 0) const override;
-  std::string toInlineString(int indent_size = 0) const override;
-};
-
-class CpAsyncBulkS2GWait final : public Expr {
+class AsyncCommit final : public Expr {
  public:
   using Expr::Expr;
 
-  explicit CpAsyncBulkS2GWait(
-      IrBuilderPasskey passkey,
-      int64_t keep_stages = 0);
+  explicit AsyncCommit(IrBuilderPasskey passkey, AsyncOpType async_op_type);
 
   NVFUSER_DECLARE_CLONE_AND_CREATE
 
   const char* getOpString() const override {
-    return "CpAsyncBulkS2GWait";
+    return "AsyncCommit";
   }
 
   std::string toString(int indent_size = 0) const override;
   std::string toInlineString(int indent_size = 0) const override;
 
-  int64_t keepStages() const {
-    return attribute<int64_t>(0);
-  }
-};
-
-class CpAsyncBulkS2GCommit final : public Expr {
- public:
-  using Expr::Expr;
-
-  explicit CpAsyncBulkS2GCommit(IrBuilderPasskey passkey);
-
-  NVFUSER_DECLARE_CLONE_AND_CREATE
+  const char* ptx() const;
+  bool memory() const;
```

Comment:
What does this function mean?

Reply:
Added a comment in the code. It basically tells if the ptx should have a `:memory` in the end. It is strange that some PTX needs it, some not....

Reply:
Side note, similar comment / documentation should be added to `AsyncWait` definition.